specification/dsdl/serializable_types.tex

```diff
@@ -0,0 +1,285 @@
+the resulting value is chosen according to the lossy assignment rules
+specified in the table \ref{table:dsdl_cast_mode}.
+Cases that are marked as illegal are not permitted in DSDL definitions.
+
+\begin{UAVCANSimpleTable}{Lossy assignment rules per cast mode}{|l X X|}
+    Type category & Truncated mode & Saturated mode (default)
+    \label{table:dsdl_cast_mode} \\
+
+    Boolean &
+    Illegal: boolean type with truncated cast mode is not allowed. &
+    Falsity if the value is zero or false, truth otherwise. \\
+
+    Signed integer &
+    Illegal: signed integer types with truncated cast mode are not allowed. &
+    Nearest reachable value. \\
+
+    Unsigned integer &
+    Most significant bits are discarded. &
+    Nearest reachable value. \\
+
+    Floating point &
+    Infinity with the same sign, unless the original value is not-a-number, in which case it will be preserved. &
+    If the original value is finite, the nearest finite value will be used.
+    Otherwise, in the case of infinity or not-a-number, the original value will be preserved. \\
+\end{UAVCANSimpleTable}
+
+Rules of conversion between values of different type categories do not affect compatibility at the protocol level,
+and as such, they are to be implementation-defined.
+
+\subsubsection{Expressions}
+
+At the time of DSDL definition processing,
+values of primitive types are represented as instances of the \verb|rational| type (section \ref{sec:dsdl_rational}),
+with the exception of the type \verb|bool|,
+instances of which are usable in DSDL expressions as-is.
+
+\begin{UAVCANSimpleTable}{Operators defined on instances of type boolean}{|l X X|}
+    Op & Type & Description \\
+
+    \texttt{\textbf{!}}     & $(\texttt{bool}) \rightarrow \texttt{bool}$ & Logical not. \\
+
+    \texttt{\textbf{||}}    & $(\texttt{bool}, \texttt{bool}) \rightarrow \texttt{bool}$ & Logical or. \\
+    \texttt{\textbf{\&\&}}  & $(\texttt{bool}, \texttt{bool}) \rightarrow \texttt{bool}$ & Logical and. \\
+
+    \texttt{\textbf{==}}    & $(\texttt{bool}, \texttt{bool}) \rightarrow \texttt{bool}$ & Equality. \\
+    \texttt{\textbf{!=}}    & $(\texttt{bool}, \texttt{bool}) \rightarrow \texttt{bool}$ & Inequality. \\
+
+\end{UAVCANSimpleTable}
+
+\subsubsection{Reference list}
+
+An exhaustive list of all void and primitive types
+ordered by bit length is provided below for reference.
+Note that the cast mode specifier is omitted intentionally.
+
+\immediate\write18{rm -f ../latex.tmp}
+\immediate\write18{../render_list_of_void_and_primitive_types.py > ../latex.tmp}
+\immediate\input{../latex.tmp}
+
+\subsection{Array types}
+
+An array type represents an ordered collection of values.
+All values in the collection share the same type, which is referred to as \emph{array element type}.
+The array element type can be any type except:
+\begin{itemize}
+    \item void type;
+    \item array type\footnote{Meaning that nested arrays are not allowed;
```

Comment:
Huh. Why not?

Reply:
We may allow them someday later. There are two weak reasons against nested arrays:

- The grammar gets slightly more complex. A naive addition of nested arrays would require us to make the array matching rule `type_array` left-recursive, which some parsers/generators may not support.  We can deal with this, though, if necessary, so it's not a huge deal.
- Variable-length nested arrays would often lead to misaligned data due to the length prefix being of weird bit width. Misaligned data is bad: https://github.com/UAVCAN/pydsdl/pull/24.

If one needs nested arrays, the following should be done instead:

```python
Nested.1.0[<=100] two_dimensional_array
```

```python
# Nested.1.0
void1  # Alignment
float16[<=100] one_dimensional_array
@assert _offset_.min % 8 == 0  # Yup, the data is aligned
```

Do you have any particular use cases in mind where nested arrays might be very desirable?